Factorize the code to make it easier to add new modes AND added a Juxtalinear edition mode

DanielC-N opened the pull request:
- added a Juxta mode that creates a translation burrito
- added a 'factoryCreateContent.js'  file to help factorize some code for 'createObsContent', 'createVersificationUSFM' and 'createAudioVersification' in 'saveProjetcsMeta.js'

- fixing some typos, fixing lint

Commit:
- added a Juxta mode that creates a translation burrito
- added a 'factoryCreateContent.js'  file to help factorize some code for 'createObsContent', 'createVersificationUSFM' and 'createAudioVersification' in 'saveProjetcsMeta.js'

- fixing some typos, fixing lint

DanielC-N commented:
First thing first I've added comments, where I did (or intend to) add code. But I've annotated it with "Nicolas" and sometimes I've added some questions with it.  
If you can answer the questions, or have a comment on my annotation or if I'm doing something wrong, please let me know!

selvabenz commented:
@DanielC-N **App is breaking, could you check? Could not move further.**

![image](https://github.com/bible-technology/scribe-scripture-editor/assets/121541505/c2bd07c0-ae99-435d-82c4-27309e06c8e8)

**CSS fix required: Icon is visible only on hover.**

![image](https://github.com/bible-technology/scribe-scripture-editor/assets/121541505/8218d87f-527e-484c-bbcc-e1be4dc031b9)

DanielC-N commented:
@selvabenz I forgot to mention that it is not yet intended to be build. "As is" it'll definitely not work since I'm still modifying the code as I'm asking questions. It's only on the code part that I need help.
Sorry for the misunderstanding.

ps : I'm terrible at css, how may I fix this Icon issue?

Commit:
Refactorization of 'saveProjectsMeta.js'

Commit:
merging changes for PR

DanielC-N commented:
@selvabenz I've updated the code so it can work properly

selvabenz commented:
Hi @DanielC-N thanks for updating the code. It is working fine. Juxta mode creates the projects under Text Translation flavour is that what you have intended? Have you added the Juxta feature in this Pr?

![image](https://github.com/bible-technology/scribe-scripture-editor/assets/121541505/96072921-9a01-427d-9bd4-bdb4e44c1058)

Commit:
Scripture Burrito as a resource : added the poissibility for the user to search and download a SB as a resource in a module

Commit:
removed checkbox from the UI and query sb resource everytime

Commit:
Added a juxtalinear editor along with a new mode.
- auto load a book on the opening of a project
- auto save after a modification
- fixed a bit of typos and code

DanielC-N commented:
@selvabenz 

I've modified the code in order to fully work with Juxta files. Now a project is correctly labeled as "Juxtalinear"  (along with the flavour)

I've some juxtalinear `JSON` files for the tests here :

[PHP.json](https://github.com/bible-technology/scribe-scripture-editor/files/14258481/PHP.json)
[TIT.json](https://github.com/bible-technology/scribe-scripture-editor/files/14258482/TIT.json)

Commit:
created a juxtaBurrito creator and a function to add juxta content. Also edited the icon for Juxta mode (it's visible now) you can create a new juxta project

Commit:
ADDED juxta support for SB creation and fixed some typos

Commit:
merged with dev

Commit:
ADDED support for usfm import for a juxta project. Now we have to change the way a juxta json is built.

Commit:
juxtalineat mode functionnal

DanielC-N committed:
Merge branch 'development' into dcn/refactor_modes

DanielC-N commented:
@selvabenz 
I've updated the code and it's working properly now. You can create a juxtalinear project by importing your own usfms. You must use greek/hebrew usfm for now.. It's correctly listed as a "Juxtalinear" flavor. When you open the project it'll open the Juxtalinear Editor.
Here's the video of the demo usage : https://youtu.be/sarZnbxbgDY

Here's some usfm files that you can import to test the mode :  
[2JN.txt](https://github.com/bible-technology/scribe-scripture-editor/files/14363062/2JN.txt)
[PHP.txt](https://github.com/bible-technology/scribe-scripture-editor/files/14363063/PHP.txt)
[TIT.txt](https://github.com/bible-technology/scribe-scripture-editor/files/14363065/TIT.txt)

DanielC-N commented:
I'm still working on the performances to load a file. For example : the book of revelation in usfm takes 10~15 sec to load for now.

Commit:
UPDATE : now the interface handle corectly a json juxta file and update it if needed. and layout color update

Commit:
function to update a json juxta file if needed

Commit:
Merge branch 'dcn/refactor_modes' of github.com:Proskomma/scribe-scripture-editor into dcn/refactor_modes

Commit:
checksums ok for each modifications of the juxta

Commit:
bugfix : juxta was still using a translation settings

Commit:
quick fix

DanielC-N committed:
Merge branch 'bible-technology:development' into development

Commit:
Merge branch 'development' into dcn/refactor_modes

Commit:
added new kewboard shortcuts : 'Tab' to navigate through line chuncks, 'double click' to modify a chunck, 'enter' to end the modification of a chunck. And added seemless navigation between Juxtalinear editor and the BibleEditorX. Navigating through verses and chapters with the menu bar will load the corresponding sentence and vice versa.

Commit:
removing a useless 'console.log'

Commit:
FIXED font size when aditing a chunck, ADDED a bit of css and added a skip for the liter for .ts files

Commit:
I put the sentences loading one layer outside the main editor, to avoid the editor to re-render everytime the juxtalinear editor module

vipinpaul commented:
Our findings, we have already shared these things with Nikola on-call

- resources collection/table took a long time to load.
- Resource is showing all versions. If needed to show all versions, make a collapsable UI for each one with the latest as default. Users can load the previous versions only by expanding the single resources.
- Academy is not showing any resources.
- Star/Unstar is not working for Juxta projects
- If the user needs the .txt or .usfm file with content for working on the Juxta editor then we shouldn't allow the user to create a Juxta project without those supporting files.
- Translation resources only show 4 resources, the rest of them are not showing.
- OBS Translation note, Questions, TWL, have missing / wrong resources are showing.
- Showing me the same content for me in all the books.
- Downloaded resources TN, TW and OBS help resources not showing in UI also in the backend.
- TWL is not loading, it keeps on spinning
- It feels like the Audio resources are reloading multiple times.
- In the project creation of JUXTA with custom scope, It is working fine in one system but another one is always taking All books
- Achieve project is not working for the text translation project, working for Juxta
 

DanielC-N commented:
* resources collection/table took a long time to load. Solved :white_check_mark: 
* Resource is showing all versions. Solved :white_check_mark: => it's only showing the latest version. Modify the code to allow the user to download old versions would be long and require some discussions
* Academy is not showing any resources. Solved :white_check_mark: 
* Star/Unstar is not working for Juxta projects. TBD 🔧
* If the user needs the .txt or .usfm file with content for working on the Juxta editor then we shouldn't allow the user to create a Juxta project without those supporting files. TBD 🔧
* Translation resources only show 4 resources, the rest of them are not showing. DONE, but needed to CHECK 🚨
* Downloaded resources TN, TW and OBS help resources not showing in UI also in the backend. DONE, but needed to CHECK 🚨
* TWL is not loading, it keeps on spinning. Solved :white_check_mark:
* It feels like the Audio resources are reloading multiple times. To discuss :question: :question: 
* In the project creation of JUXTA with custom scope, It is working fine in one system but another one is always taking All books. To discuss :question: :question: 
* Achieve project is not working for the text translation project, working for Juxta. To discuss :question: :question:

Commit:
Many bugfixes https://github.com/bible-technology/scribe-scripture-editor/pull/303#issuecomment-2034384825

DanielC-N commented:
Bonus : https://git.door43.org/ru_gl/ru_obs-tn

I've fixed a bug here : https://github.com/Proskomma/scribe-scripture-editor/blob/7eafc20c7dc87c89b4d8f88fcd31f9dbdf1b4653/renderer/src/components/EditorPage/Reference/OBS/ObsTn.js#L86

The problem show up when the user put the path to the FOLDER instead of the name of the tsv file in `metadata.projects[0].path`

Commit:
add a check if the folder exists

Commit:
added a little bit of secure code for OBS loading

Commit:
changed the behavior of the Snackbar. If error === 'failure' the snackbar will show up for a longer time

Commit:
changed the behavior of the Snackbar. If error === 'failure' the snackbar will show up for a longer time

Commit:
bold text 'CREATE PROJECT' for better visibility and start to check 'canonSpecification' for a juxta project

Commit:
added the feature that download resources not imported by the user for juxtalinear project

Commit:
Now the user is correctly notified when the imported books does not match the canon specification, and also, the canon specification automatically matches the file imported

DanielC-N commented:
* resources collection/table took a long time to load. Solved :white_check_mark: 
* Resource is showing all versions. Solved :white_check_mark: => it's only showing the latest version. Modify the code to allow the user to download old versions would be long and require some discussions
* Academy is not showing any resources. Solved :white_check_mark: 
* Star/Unstar is not working for Juxta projects. TBD 🔧
* If the user needs the .txt or .usfm file with content for working on the Juxta editor then we shouldn't allow the user to create a Juxta project without those supporting files. DONE e554082 :white_check_mark:
* Translation resources only show 4 resources, the rest of them are not showing. DONE :white_check_mark: , but needed to CHECK 🚨
* Downloaded resources TN, TW and OBS help resources not showing in UI also in the backend. DONE :white_check_mark: , but needed to CHECK 🚨
* TWL is not loading, it keeps on spinning. Solved :white_check_mark:
* It feels like the Audio resources are reloading multiple times. To discuss :question: :question: 
* In the project creation of JUXTA with custom scope, It is working fine in one system but another one is always taking All books. To discuss :question: :question: 
* Archive project is not working for the text translation project, working for Juxta. To discuss :question: :question:

Commit:
minor bugfixes and code cleaning

Commit:
changed place of the 'nprogress' by few px

selvabenz commented:
@DanielC-N 
- Unable to create a new project with JUSTA mode on. 
![image](https://github.com/bible-technology/scribe-scripture-editor/assets/121541505/e93bc8d0-b325-4d68-a76e-fa860efb7985)

Commit:
added a better user-friendly messages to nugde the user to import at least one resource for a Juxtalinear project

Commit:
FIXED : archive a Text Translation project OK.

DanielC-N commented:
@selvabenz I've updated the messages given to the user! It's clearer now.

DanielC-N commented:
* resources collection/table took a long time to load. Solved :white_check_mark: 
* Resource is showing all versions. Solved :white_check_mark: => it's only showing the latest version. Modify the code to allow the user to download old versions would be long and require some discussions
* Academy is not showing any resources. Solved :white_check_mark: 
* Star/Unstar is not working for Juxta projects. TBD 🔧
* If the user needs the .txt or .usfm file with content for working on the Juxta editor then we shouldn't allow the user to create a Juxta project without those supporting files. DONE e554082 :white_check_mark:
* Translation resources only show 4 resources, the rest of them are not showing. DONE :white_check_mark: , but needed to CHECK 🚨
* Downloaded resources TN, TW and OBS help resources not showing in UI also in the backend. DONE :white_check_mark: , but needed to CHECK 🚨
* TWL is not loading, it keeps on spinning. Solved :white_check_mark:
* It feels like the Audio resources are reloading multiple times. To discuss :question: :question: 
* In the project creation of JUXTA with custom scope, It is working fine in one system but another one is always taking All books. To discuss :question: :question: 
* Archive project is not working for the text translation project, working for Juxta. Solved :white_check_mark: (a0f4b91)

Commit:
star/unstar a juxtalinear project works now

DanielC-N commented:
* resources collection/table took a long time to load. Solved :white_check_mark: 
* Resource is showing all versions. Solved :white_check_mark: => it's only showing the latest version. Modify the code to allow the user to download old versions would be long and require some discussions
* Academy is not showing any resources. Solved :white_check_mark: 
* **Star/Unstar is not working for Juxta projects. Done :white_check_mark:** ([58605c8](https://github.com/bible-technology/scribe-scripture-editor/pull/303/commits/58605c83923bf8b40dc29fc5d9182358b0c51ee2))
* If the user needs the .txt or .usfm file with content for working on the Juxta editor then we shouldn't allow the user to create a Juxta project without those supporting files. DONE e554082 :white_check_mark:
* Translation resources only show 4 resources, the rest of them are not showing. DONE :white_check_mark: , but needed to CHECK 🚨
* Downloaded resources TN, TW and OBS help resources not showing in UI also in the backend. DONE :white_check_mark: , but needed to CHECK 🚨
* TWL is not loading, it keeps on spinning. Solved :white_check_mark:
* It feels like the Audio resources are reloading multiple times. To discuss :question: :question: 
* In the project creation of JUXTA with custom scope, It is working fine in one system but another one is always taking All books. To discuss :question: :question: 
* Archive project is not working for the text translation project, working for Juxta. Solved :white_check_mark: (a0f4b91)

Commit:
FIX : bug that prevented some resources to show up. and FIX : bug that was throwing an error when a user where to search for a resource in the resource search field, and click on 'no match found'

sijumoncy commented on package-lock.json:
We moved away from npm and currently using yarn. can you remove the package-lock from the PR ? Only need to keep yarn lock

Commit:
deleted package-lock.json and updated yarn.lock

DanielC-N committed:
Merge branch 'bible-technology:development' into development

Commit:
resolved conficts

Commit:
updated french translation

Commit:
bugfix : you can now edit a juxta project properly

Commit:
fixed bug where on book change, the sentences where wronly saved under another filename

DanielC-N commented:
@selvabenz 
I've fixed the bug I've found in my modifications => e5fd90b

You can test it yourself by creating a project, with 2 different files, then navigating between the two books in the project. Now the sentences are correctly loaded.

DanielC-N committed:
Merge branch 'development' into dcn/refactor_modes

Commit:
make 'jsonFileContent.bookCode' to upper case for safety reasons

DanielC-N committed:
Merge branch 'development' into dcn/refactor_modes

selvabenz commented:
Unable to export Juxta projects, the app is frozen during the export. All the other flavor exports well.

![image](https://github.com/bible-technology/scribe-scripture-editor/assets/121541505/b25fc700-85e7-4821-a7be-013f6d50d90b)

Commit:
FIX : export/import now works properly for juxtalinear projects

Commit:
Merge branch 'dcn/refactor_modes' of github.com:Proskomma/scribe-scripture-editor into dcn/refactor_modes

DanielC-N commented:
@selvabenz FIX for the export (and import) problem : b4d8e3c
It now works properly.

selvabenz commented:
> @selvabenz FIX for the export (and import) problem : [b4d8e3c](https://github.com/bible-technology/scribe-scripture-editor/commit/b4d8e3c43cb5ec81090d0ee3d96a0e656a345945) It now works properly.

Will check and respond soon @DanielC-N

selvabenz commented:
@DanielC-N i am facing a problem in the OBS mode, while the project is exported there is no project as zip checkbox visible but it is exporting the project as a zip file. 

Note: In the dev branch project as zip only assigned to text translation

Commit:
fix obs auto zipping on export

DanielC-N commented:
@selvabenz I've fixed the problem here : 3dda2e5

I did modify this behavior without noticing the change for obs. In the long run, I think this part of the code could use some refactorisation.

selvabenz commented:
@DanielC-N Sounds good! thanks for the fix.

@vipinpaul they did not do the CSS fix on the create project button and the spinner @joelthe1 this PR can be merged.

Commit:
spinner and 'create project' button css fix.

Commit:
'create project' button reverted

selvabenz commented:
@vipinpaul and @joelthe1 this PR can be merged. all the asked CSS fix have been made.

DanielC-N committed:
Merge branch 'bible-technology:development' into development

Commit:
resolving conflict merge, it was due to a renaming of a variable to avoid confusions
